### ISSUE TITLE
# "kics-scan ignore-line" comment doesn't work in YAML to ignore secret false positive in next commented line

### ISSUE THREAD

**Issue by HariSekhon** [Checkmarx/kics#5177]
### Expected Behavior

Ignore a false positive line when preceeded by a line such as:
```yaml
# kics-scan ignore-line
```

Iirc this was supposed to be fixed in PR #4662 to solve #4563 and #4661. I'm not sure if this is because the false positive found token is in a commented line instead of a live yaml line, but it still results in a false positive and not ignoring the ignore-line.

### Actual Behavior

Still results in the false positive - tested in versions 1.5.0, 1.5.1 and 1.5.2 in CLI, Docker Image and GitHub Actions checkmarx provided action.

### Steps to Reproduce the Problem

Live GitHub Actions workflow example can be seen here:

https://github.com/HariSekhon/GitHub-Actions/runs/5976076251?check_suite_focus=true#step:8:52

Docker Reproduction:

```shell
$ git clone git@github.com:HariSekhon/GitHub-Actions.git
$ cd GitHub-Actions/
$ docker run -v $PWD:/path checkmarx/kics scan -p /path -o /path

                   .0MO.                                    
                   OMMMx                                    
                   ;NMX;                                    
                    ...           ...              ....     
WMMMd     cWMMM0.  KMMMO      ;xKWMMMMNOc.     ,xXMMMMMWXkc.
WMMMd   .0MMMN:    KMMMO    :XMMMMMMMMMMMWl   xMMMMMWMMMMMMl
WMMMd  lWMMMO.     KMMMO   xMMMMKc...'lXMk   ,MMMMx   .;dXx 
WMMMd.0MMMX;       KMMMO  cMMMMd        '    'MMMMNl'       
WMMMNWMMMMl        KMMMO  0MMMN               oMMMMMMMXkl.  
WMMMMMMMMMMo       KMMMO  0MMMX                .ckKWMMMMMM0.
WMMMMWokMMMMk      KMMMO  oMMMMc              .     .:OMMMM0
WMMMK.  dMMMM0.    KMMMO   KMMMMx'    ,kNc   :WOc.    .NMMMX
WMMMd    cWMMMX.   KMMMO    kMMMMMWXNMMMMMd .WMMMMWKO0NMMMMl
WMMMd     ,NMMMN,  KMMMO     'xNMMMMMMMNx,   .l0WMMMMMMMWk, 
xkkk:      ,kkkkx  okkkl        ;xKXKx;          ;dOKKkc    


Scanning with Keeping Infrastructure as Code Secure v1.5.5


Preparing Scan Assets:  \Preparing Scan Assets:  -Preparing Scan Assets:  |Preparing Scan Assets:  /Preparing Scan Assets:  -Preparing Scan Assets:  |Preparing Scan Assets:  \Preparing Scan Assets:  -Preparing Scan Assets:  |Preparing Scan Assets:  /Preparing Scan Assets:  -Preparing Scan Assets:  |Preparing Scan Assets:  \Preparing Scan Assets:  -Preparing Scan Assets:  |Preparing Scan Assets:  /Preparing Scan Assets:  -Preparing Scan Assets:  |Preparing Scan Assets:  \Preparing Scan Assets:  -Preparing Scan Assets:  |Preparing Scan Assets:  /Preparing Scan Assets:  -Preparing Scan Assets:  |Preparing Scan Assets:  \Preparing Scan Assets:  -Preparing Scan Assets:  |Preparing Scan Assets:  /Preparing Scan Assets:  -Preparing Scan Assets:  |Preparing Scan Assets:  \Preparing Scan Assets:  -Preparing Scan Assets:  |Preparing Scan Assets:  /Preparing Scan Assets:  -Preparing Scan Assets:  |Preparing Scan Assets:  \Preparing Scan Assets:  -Preparing Scan Assets:  |Preparing Scan Assets:  /Preparing Scan Assets:  -Preparing Scan Assets:  |Preparing Scan Assets:  \Preparing Scan Assets:  -Preparing Scan Assets:  |Preparing Scan Assets:  /Preparing Scan Assets:  -Preparing Scan Assets:  |Preparing Scan Assets:  \Preparing Scan Assets:  -Preparing Scan Assets:  |Preparing Scan Assets:  /Preparing Scan Assets:  -Preparing Scan Assets:  |Preparing Scan Assets:  \Preparing Scan Assets:  -Preparing Scan Assets:  |Preparing Scan Assets:  /Preparing Scan Assets:  -Preparing Scan Assets:  |Preparing Scan Assets:  \Preparing Scan Assets:  -Preparing Scan Assets:  |Preparing Scan Assets:  /Preparing Scan Assets:  -Preparing Scan Assets:  |Preparing Scan Assets:  \Preparing Scan Assets:  -Preparing Scan Assets:  |Preparing Scan Assets:  /Preparing Scan Assets:  -Preparing Scan Assets:  |Preparing Scan Assets:  \Preparing Scan Assets:  -Preparing Scan Assets:  |Preparing Scan Assets:  /Preparing Scan Assets:  -Preparing Scan Assets:  |Preparing Scan Assets:  \Preparing Scan Assets:  -Preparing Scan Assets: DoneExecuting queries: [>________________________________________________________________________] 0.39%Executing queries: [>________________________________________________________________________] 0.78%Executing queries: [->_______________________________________________________________________] 1.55%Executing queries: [->_______________________________________________________________________] 1.94%Executing queries: [->_______________________________________________________________________] 2.33%Executing queries: [->_______________________________________________________________________] 2.71%Executing queries: [-->______________________________________________________________________] 3.49%Executing queries: [-->______________________________________________________________________] 3.88%Executing queries: [--->_____________________________________________________________________] 4.26%Executing queries: [--->_____________________________________________________________________] 4.65%Executing queries: [--->_____________________________________________________________________] 5.04%Executing queries: [--->_____________________________________________________________________] 5.43%Executing queries: [---->____________________________________________________________________] 6.20%Executing queries: [---->____________________________________________________________________] 6.59%Executing queries: [----->___________________________________________________________________] 6.98%Executing queries: [----->___________________________________________________________________] 7.36%Executing queries: [----->___________________________________________________________________] 8.14%Executing queries: [------>__________________________________________________________________] 8.53%Executing queries: [------>__________________________________________________________________] 8.91%Executing queries: [------>__________________________________________________________________] 9.30%Executing queries: [------->_________________________________________________________________] 9.69%Executing queries: [------->________________________________________________________________] 10.08%Executing queries: [------->________________________________________________________________] 10.47%Executing queries: [------->________________________________________________________________] 10.85%Executing queries: [-------->_______________________________________________________________] 11.24%Executing queries: [-------->_______________________________________________________________] 11.63%Executing queries: [-------->_______________________________________________________________] 12.02%Executing queries: [------------->__________________________________________________________] 19.38%Executing queries: [--------------------->__________________________________________________] 29.84%Executing queries: [--------------------------->____________________________________________] 38.37%Executing queries: [-------------------------------->_______________________________________] 45.74%Executing queries: [------------------------------------>___________________________________] 50.39%Executing queries: [-------------------------------------->_________________________________] 53.10%Executing queries: [----------------------------------------->______________________________] 57.36%Executing queries: [------------------------------------------------>_______________________] 67.44%Executing queries: [------------------------------------------------------->________________] 77.13%Executing queries: [-------------------------------------------------------------->_________] 87.21%Executing queries: [--------------------------------------------------------------------->__] 96.12%Executing queries: [----------------------------------------------------------------------->] 99.61%Executing queries: [-----------------------------------------------------------------------] 100.00%
Files scanned: 41
Parsed files: 41
Queries loaded: 288
Queries failed to execute: 0

------------------------------------

Passwords And Secrets - Generic Token, Severity: HIGH, Results: 1
Description: Query to find passwords and secrets in infrastructure code.
Platform: Common

        [1]: ../../path/main.yaml:226

                225:   # kics-scan ignore-line
                226:   #id-token: read|write|none
                227:   #issues: read|write|none



Results Summary:
HIGH: 1
MEDIUM: 0
LOW: 0
INFO: 0
TOTAL: 1

Results saved to file /path/results.json
Generating Reports: DoneScan duration: 27.922168s
```

### Specifications

-   Version: 1.5.0, 1.5.1, 1.5.2
-   Platform: Linux, macOS

**Comment by rafaela-soares** [comment]
Hello, @HariSekhon! Thank you so much for contacting us! 😊 

Can you provide the file to us, please?

P.S. I tried the following test cases, and it does not return the result:

```
- name: Start a workflow in the Itential Automation Platform
  community.network.iap_start_workflow:
      iap_port: 3000
      iap_fqdn: localhost
      #id-token: read|write|none
      #issues: read|write|none
      workflow_name: "RouterUpgradeWorkflow"
      description: "OS-Router-Upgrade"
      variables: {"deviceName":"ASR9K"}
```

```
- name: Start a workflow in the Itential Automation Platform
  community.network.iap_start_workflow:
      iap_port: 3000
      iap_fqdn: localhost
      # kics-scan ignore-line
      #id-token: read|write|none
      #issues: read|write|none
      workflow_name: "RouterUpgradeWorkflow"
      description: "OS-Router-Upgrade"
      variables: {"deviceName":"ASR9K"}
```

**Comment by HariSekhon** [comment]
Thanks for confirming - I'm glad it's not just me then.

I look forward to the fix so my kics badges can go green again!

**Comment by rafaela-soares** [comment]
As a workaround, you can use the flag `--exclude-results <similarity_id>`. Something like: `--exclude-results 01271c53e0ed42b21000a92fd926a473beac1ec98bde049e301dfdae84e5d01a`

You can find the similarity id of the result in the JSON report:

![image](https://user-images.githubusercontent.com/74001161/162948367-f72f0e81-439f-4e20-a0f7-c2d6afab4a45.png)

**Comment by rafaela-soares** [comment]
Hello, @HariSekhon!

I am sending this message to let you know that we did not forget about this issue.

Indeed, your file found a few problems in our comment approach. Thank you so much for reporting it.

### Problem 1
One of them is related to how KICS interprets the content of the comments. KICS uses `gopkg.in/yaml.v3` to get information about the comments of the YAML nodes (through HeadComment, FootComment).
For now, KICS counts how many `#` is in the "content" of the comment and then KICS tries to get the lines to ignore by considering this number, as can be seen [here](https://github.com/Checkmarx/kics/blob/master/pkg/model/comment_yaml.go#L124:L139). This approach has its limitations when the content of the comment has empty lines or has `#` in the comment line itself. 

As an example, take a look at how the node is presented (focus only on HeadComment please. I do not know why the node presents a FootCooment. I opened a [issue](https://github.com/go-yaml/yaml/issues/846) on their side to understand):
![image](https://user-images.githubusercontent.com/74001161/165491949-6cad4881-6559-4ac2-9f92-7504bfeb3a0c.png)

KICS will count the number of `#` between lines 1 and 20 (it is 20). KICS will consider that the 20 lines before line 24 are comment lines. So, KICS gets the lines 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, and 23.

![image](https://user-images.githubusercontent.com/74001161/165505471-405f1f25-22fb-45bf-8805-782e9b4c7213.png)

This approach will be reviewed. However, it also depends on the [issue response](https://github.com/go-yaml/yaml/issues/846), since there is a "strange" behavior in how the new lines are interpreted.


### Problem 2
The FootComment [in this particular case](https://github.com/HariSekhon/GitHub-Actions/blob/master/main.yaml#L221:L234) is not being considered by KICS and we will take a look as soon as possible.


This issue can take some time to close. We apologize in advance.

**Comment by HariSekhon** [comment]
Thanks for the update, I appreciate that you're working on it.

**Comment by HariSekhon** [comment]
Any update on this issue? It's been over a year and I'm still hitting this...

**Comment by vparmeland** [comment]
Created in 2022 and last update in 2023... up please
